Added Client Message Events

kevinthegreat1 opened the pull request:
Added Client Message Events including

- AllowSendChatMessage, SendChatMessage
- AllowSendCommandMessage, SendCommandMessage
- AllowReceiveChatMessage, ReceiveChatMessage
- AllowReceiveGameMessage, ReceiveGameMessage

kevinthegreat1 committed:
Added Client Message Events

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/message/MessageHandlerMixin.java:
```diff
@@ -0,0 +1,62 @@
+public abstract class MessageHandlerMixin {
+
+	@Inject(method = "processChatMessageInternal", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/hud/ChatHud;addMessage(Lnet/minecraft/text/Text;Lnet/minecraft/network/message/MessageSignatureData;Lnet/minecraft/client/gui/hud/MessageIndicator;)V", ordinal = 0), cancellable = true)
+	private void fabric_message_api_onSignedChatMessage(MessageType.Parameters params, SignedMessage message, Text decorated, GameProfile sender, boolean onlyShowSecureChat, Instant receptionTimestamp, CallbackInfoReturnable<Boolean> cir) {
```
No need to prefix the inject method names with `fabric_message_api_`

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/message/MessageHandlerMixin.java:
```diff
@@ -0,0 +1,62 @@
+		fabric_message_api_onChatMessage(decorated, null, null, params, receptionTimestamp, cir);
+	}
+
+	private void fabric_message_api_onChatMessage(Text message, @Nullable SignedMessage signedMessage, @Nullable GameProfile sender, MessageType.Parameters params, Instant receptionTimestamp, CallbackInfoReturnable<Boolean> cir) {
```
Just prefix with fabric_, I think you can also put @Unique on it

modmuss50 commented on fabric-message-api-v1/src/client/resources/fabric-message-api-v1.client.mixins.json:
```diff
@@ -0,0 +1,13 @@
+  "required": true,
+  "package": "net.fabricmc.fabric.mixin.message",
+  "compatibilityLevel": "JAVA_16",
+  "mixins": [
+  ],
```
Nit: Remove

modmuss50 commented on fabric-message-api-v1/src/client/resources/fabric-message-api-v1.client.mixins.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "required": true,
+  "package": "net.fabricmc.fabric.mixin.message",
+  "compatibilityLevel": "JAVA_16",
```
Nit: Might as well use JAVA_17

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		boolean allowSendChatMessage(String message);
+	}
+
+	public interface AllowSendCommandMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		boolean allowSendCommandMessage(String message);
+	}
+
+	public interface SendChatMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		void onSendChatMessage(String message);
+	}
+
+	public interface SendCommandMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		void onSendCommandMessage(String message);
+	}
+
+	public interface AllowReceiveChatMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		boolean allowReceiveChatMessage(Text message, @Nullable SignedMessage signedMessage, @Nullable GameProfile sender, MessageType.Parameters params, Instant receptionTimestamp);
+	}
+
+	public interface AllowReceiveGameMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		boolean allowReceiveGameMessage(Text message, boolean overlay);
+	}
+
+	public interface ReceiveChatMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+		void onReceiveChatMessage(Text message, @Nullable SignedMessage signedMessage, @Nullable GameProfile sender, MessageType.Parameters params, Instant receptionTimestamp);
+	}
+
+	public interface ReceiveGameMessage {
```
Missing @FunctionalInterface

modmuss50 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,256 @@
+package net.fabricmc.fabric.api.message.v1;
```
Missing package header, I see a few checkstyle issues as well. Please make sure this builds.

kevinthegreat1 committed:
Applied suggestions and fixed checkstyle

apple502j commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,283 @@
+/**
+ * Contains client-side events triggered when sending or receiving messages.
+ */
+public class ClientMessageEvents {
```
Make it final, also add private constructor

apple502j commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,283 @@
+	});
+
+	/**
+	 * An event triggered when the client received a chat message,
```
"receives"?

apple502j commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,283 @@
+	});
+
+	/**
+	 * An event triggered when the client received a chat message,
```
Same

apple502j commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,283 @@
+	@FunctionalInterface
+	public interface AllowSendCommandMessage {
+		/**
+		 * An event triggered when the client is about to send a command,
```
Called...

apple502j commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/message/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,49 @@
+
+import net.fabricmc.fabric.api.message.v1.ClientMessageEvents;
+
+@Mixin(ClientPlayNetworkHandler.class)
```
Should set priority to make sure client commands don't get handled

kevinthegreat1 committed:
Inject before fabric-command-api and updated Javadocs

kevinthegreat1 commented:
Ok, should be fixed. Also [this](https://github.com/FabricMC/fabric/pull/2618) should be ready for merge.

apple502j commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/message/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,53 @@
+ * Mixin to {@link ClientPlayNetworkHandler} to listen for sending messages and commands.
+ * Priority set to 800 to inject before {@code fabric-command-api} so that this api will be called first.
+ */
+@Mixin(value = ClientPlayNetworkHandler.class, priority = 800)
```
I mean, run it after client commands (the docs say "server-side command")

kevinthegreat1 committed:
Updated Javadocs regarding client commands

kevinthegreat1 commented:
I think this event should capture all messages and commands sent/executed by the player, including client commands. This gives a way to override all messages and commands. I have updated the docs to reflect this behavior.

Juuxel commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,301 @@
+		 * Therefore, client commands registered with {@code fabric-command-api}
+		 * will not be executed if {@code false} is returned.
+		 *
+		 * @param message the command that will be sent to the server
```
Does this have the initial slash? I feel that's a good thing to document somewhere

Juuxel commented on fabric-message-api-v1/src/client/resources/fabric-message-api-v1.client.mixins.json:
```diff
@@ -0,0 +1,11 @@
+  "package": "net.fabricmc.fabric.mixin.message",
+  "compatibilityLevel": "JAVA_17",
+  "client": [
+    "ClientPlayNetworkHandlerMixin", "MessageHandlerMixin"
```
```suggestion
    "ClientPlayNetworkHandlerMixin",
    "MessageHandlerMixin"
```

kevinthegreat1 committed:
Update fabric-message-api-v1/src/client/resources/fabric-message-api-v1.client.mixins.json

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

kevinthegreat1 committed:
Updated Javadocs regarding commands

kevinthegreat1 commented:
I applied the suggestions.

apple502j commented:
@kevinthegreat1 Can you test if the PR still works with the latest prerelease?

kevinthegreat1 committed:
Merge branch '1.19.3' into clientScreenEvents

kevinthegreat1 committed:
Fixed duplicated package names

kevinthegreat1 committed:
Merge branch '1.19.3' into clientScreenEvents

kevinthegreat1 commented:
@apple502j Still works in 1.19.3-rc1. I don't expect them to make changes to chat for at least a while.

kevinthegreat1 commented:
I don't know why the build failed. Also, I can't find a rerun button in the details now. Can somebody rerun the checks?

Sideroo commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/message/MessageHandlerMixin.java:
```diff
@@ -0,0 +1,82 @@
+
+	@Unique
+	private void fabric_onChatMessage(Text message, @Nullable SignedMessage signedMessage, @Nullable GameProfile sender, MessageType.Parameters params, Instant receptionTimestamp, CallbackInfoReturnable<Boolean> cir) {
+		if (!ClientMessageEvents.ALLOW_RECEIVE_CHAT_MESSAGE.invoker().allowReceiveChatMessage(message, signedMessage, sender, params, receptionTimestamp)) {
+			cir.setReturnValue(false);
+			return;
+		}
```
Should log the message and set the last process time regardless of the event

kevinthegreat1 replied:
The chat log is only used for chat reporting. If the player doesn't see the chat message, should we still add it to the log? I think this needs more discussion.

Sideroo replied:
I thought logging every possible message was minecraft's default behavior but there isn't a case where message isn't sent. So it only logs the sent messages. You are right, I don't think it's neccessary.

Sideroo commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/message/MessageHandlerMixin.java:
```diff
@@ -0,0 +1,82 @@
+
+	@Inject(method = "onGameMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/hud/ChatHud;addMessage(Lnet/minecraft/text/Text;)V"), cancellable = true)
+	private void fabric_onGameMessage(Text message, boolean overlay, CallbackInfo ci) {
+		if (!ClientMessageEvents.ALLOW_RECEIVE_GAME_MESSAGE.invoker().allowReceiveGameMessage(message, overlay)) {
+			ci.cancel();
+			return;
+		}
```
Should log the message regardless of the event.

Sideroo commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/message/MessageHandlerMixin.java:
```diff
@@ -0,0 +1,82 @@
+
+	@Inject(method = "onGameMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/hud/InGameHud;setOverlayMessage(Lnet/minecraft/text/Text;Z)V"), cancellable = true)
+	private void fabric_onOverlayGameMessage(Text message, boolean overlay, CallbackInfo ci) {
+		fabric_onGameMessage(message, overlay, ci);
```
Message is not logged if it's an overlay message. So also change this when you update the pr.

Sideroo reviewed (changes requested):
Besides the javadoc changes:
* Don't immediately set the return value when `ALLOW_RECIEVE_CHAT_MESSAGE` returns false, setting the return value ends the method, before minecraft logs the message to it's internal log for security purposes and sets the last process time stamp. Fabric should handle these if event returns false.
* First one applies to `ALLOW_RECIEVE_GAME_MESSAGE` as well.

Also `CANCELLED` events would be nice. 
Or invoke `RECEIVE` events regardless of `ALLOW` events and pass the result of the allow event to the receive event for context.

Sideroo reviewed (changes requested):
Besides the javadoc changes:
* Don't immediately set the return value when `ALLOW_RECIEVE_CHAT_MESSAGE` returns false, setting the return value ends the method, before minecraft logs the message to it's internal log for security purposes and sets the last process time stamp. Fabric should handle these if event returns false.
* First one applies to `ALLOW_RECIEVE_GAME_MESSAGE` as well.

Also `CANCELLED` events would be nice. 
Or invoke `RECEIVE` events regardless of `ALLOW` events and pass the result of the allow event to the receive event for context.

kevinthegreat1 commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/client/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,305 @@
+		 * Is not called when {@linkplain #ALLOW_SEND_COMMAND_MESSAGE command messages are blocked}.
+		 * The command string does not include a slash at the beginning.
+		 *
+		 * <p>This will also be called when executing a client command
+		 * registered with {@code fabric-command-api}.
```
I don't see a problem with having this twice. The whole point is to have the Javadoc for the event and the interface similar, so looking at either Javadoc will give you the same information. This is also the case in the existing `ServerMessageEvents`.

Sideroo replied:
I didn't mean the event. I was referring to line 225, which says that the event includes client commands. Then it is repeated here, just phrased differently.

kevinthegreat1 replied:
Sorry, I misunderstood you. I removed the duplicate.

kevinthegreat1 committed:
Updated ClientMessageEvents.java Javadoc

Co-authored-by: Sideroo <109681866+Sideroo@users.noreply.github.com>

Sideroo commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/client/message/v1/ClientMessageEvents.java:
```diff
@@ -0,0 +1,305 @@
+		 * the {@link #SEND_COMMAND_MESSAGE} event from triggering.
+		 * The command string does not include a slash at the beginning.
+		 *
+		 * <p>Also, this is called before client commands are checked and executed.
+		 * Therefore, client commands registered with {@code fabric-command-api}
+		 * will not be executed if {@code false} is returned.
```
Same here, referring to line 193.

kevinthegreat1 committed:
Removed duplicated client commands Javadoc

kevinthegreat1 committed:
Added cancelled sending and receiving events

kevinthegreat1 committed:
Merge branch '1.19.3' into clientScreenEvents

kevinthegreat1 committed:
Seperated send and receive events and changed event names

kevinthegreat1 committed:
Fixed checkstyle

Technici4n commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/message/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * Mixin to {@link ClientPlayNetworkHandler} to listen for sending messages and commands.
+ * Priority set to 800 to inject before {@code fabric-command-api} so that this api will be called first.
```
Did you test this specifically? Might need a higher priority cause we want our `HEAD` to be applied last wrt. the other `HEAD` injects... needs to be confirmed by testing.

kevinthegreat1 replied:
I'm not sure what you mean. What is wrong with having this mixin applied first?

Technici4n replied:
If my understanding is correct:

```
after first mixin:

<-- method start
<-- HEAD inject at priority 800
<-- method body

after second mixin

<-- method start
<-- HEAD inject at priority 1000
<-- rest of the method (i.e. previous inject and then body)
```

So to be sure I recommend checking the generated classes using `-Dmixin.debug.export=true`.

kevinthegreat1 replied:
The HEAD inject with priority 800 (allowSendCommandMessage) is executed first.
<img width="1233" alt="Screenshot 2022-12-25 at 11 57 09" src="https://user-images.githubusercontent.com/92656833/209480529-2a4579e4-cb78-449a-96d3-a43afc57aed3.png">

Technici4n replied:
All good then.

Technici4n commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/message/MessageHandlerMixin.java:
```diff
@@ -0,0 +1,82 @@
+		}
+	}
+
+	@Inject(method = "onGameMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/hud/InGameHud;setOverlayMessage(Lnet/minecraft/text/Text;Z)V"), cancellable = true)
+	private void fabric_onOverlayGameMessage(Text message, boolean overlay, CallbackInfo ci) {
+		fabric_onGameMessage(message, overlay, ci);
+	}
+
+	@Inject(method = "onGameMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/hud/ChatHud;addMessage(Lnet/minecraft/text/Text;)V"), cancellable = true)
+	private void fabric_onGameMessage(Text message, boolean overlay, CallbackInfo ci) {
```
You should be able to use a `JUMP` injection point to inject before the `if (overlay)`. Not sure if it's worth it though. Otherwise you can specify two selectors in the `at` to only have a single method.

Technici4n reviewed (changes requested):
Would it be useful to have a way to modify sent or received messages? The notification events aren't doing much atm.

kevinthegreat1 commented:
Modifying a received chat message seems to be difficult. The signature and narration would be wrong if we change the message. Is it ok if we only changed the displayed message and leave the narration? Also, the player might be seeing different messages than the ones in the log which causes all sorts of issues with chat reporting.

kevinthegreat1 committed:
Added support for modifying messages

Technici4n commented:
> Also, the player might be seeing different messages than the ones in the log which causes all sorts of issues with chat reporting.

Ah right, that's a good point.

kevinthegreat1 committed:
Added client command test

kevinthegreat1 committed:
Added narration and message indicator support for modifying received messages

kevinthegreat1 committed:
Added tests for modifying messages

kevinthegreat1 commented:
I have added the ability to modify messages. Currently, modified received chat messages are marked with the `Modified by Server` indicator and modified received game messages are not changed and mark with `Game Message` indicator. This is not really accurate and to make it accurate we will have to add our own indicator like `Chat/Game Message Modified by client`.

kevinthegreat1 commented:
I have not tested this on multiplayer servers. How do I build a jar for the test mod?

Technici4n commented:
> I have not tested this on multiplayer servers. How do I build a jar for the test mod?

`gradlew remapTestmodJar` I think?

Technici4n reviewed (changes requested):
The modification events should have modify in the name I think. So `GAME` -> `MODIFY_GAME`, `Game` -> `ModifyGame`, `onReceiveGameMessage` -> `modifyReceivedGameMessage` or something like that.

Do we need plain notification events that get called with the modified message? (Don't think so).

apple502j reviewed (changes requested):
Modification should internally cancel the message (like when Social Interactions hides the message). Otherwise this can mess up message chain.

Technici4n commented:
In that case, maybe it's best to leave out modification entirely (for now at least) ?

kevinthegreat1 commented:
Currently, it should keep the original message in the chain. When reporting, you should see the unmodified message.

kevinthegreat1 commented:
> The modification events should have modify in the name I think. So `GAME` -> `MODIFY_GAME`, `Game` -> `ModifyGame`, `onReceiveGameMessage` -> `modifyReceivedGameMessage` or something like that.
> 
> Do we need plain notification events that get called with the modified message? (Don't think so).

The main problem is that people won't know which one to use if they just want to listen for messages. Thats why I didn't add `MODIFY` in front.

kevinthegreat1 committed:
Updated ClientReceiveMessageEvents#CHAT Javadocs

kevinthegreat1 commented:
Works as intended in multiplayer. Modified secure chat messages are marked with modified indicator when modified on receive. Already modified messages or game messages are not changed. Chat reports use original unmodified and signed messages.

apple502j reviewed (changes requested):
Please, either remove modification, or cancel the message. Having a "modified" message show up in hud while being logged in chat log as unmodified is very bad.

(by "cancelling" I mean early-returning `false` in `processChatMessageInternal`)

kevinthegreat1 commented:
Just to be clear: the unmodified and original message is logged in the chat log, not the modified message.

Technici4n commented:
Yeah, that's quite problematic indeed. I think modification should be removed entirely. 😓

kevinthegreat1 commented:
I think if we remove it, mod devs would use less thoughtful ways to achieve this and cause more trouble for users.

kevinthegreat1 commented:
I don't see a big problem with the current implementation. Players see the original unmodified message when hovering over the message and when reporting.

apple502j commented:
@kevinthegreat1 I believe the best solution is, as I said, to cancel and resend. If you're busy I can open a PR with that. What do you think?

kevinthegreat1 committed:
Small Javadoc fixes

kevinthegreat1 commented:
I think the best solution is what we have currently. There will be gaps in the log and lack of context when reporting if we cancel the modified messages. Players see the original message when reporting. If you really don't think this is good I can update the pr.

Technici4n commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/client/message/v1/ClientReceiveMessageEvents.java:
```diff
@@ -0,0 +1,229 @@
+	 * Messages already with the modified indicator or game messages are not affected.
+	 * The unmodified message is used in chat reporting.
+	 */
+	public static final Event<Chat> CHAT = EventFactory.createArrayBacked(Chat.class, listeners -> (message, signedMessage, sender, params, receptionTimestamp) -> {
```
Should probably be `MODIFY_CHAT` since the message can be modified. (and the function name would be `modifyReceivedChatMessage`).

kevinthegreat1 committed:
Added Modify to names

kevinthegreat1 committed:
Merge branch '1.19.3' into clientScreenEvents

apple502j reviewed (changes requested):
I think this breaks custom message types with different narrations.

kevinthegreat1 committed:
Always narrate original message

kevinthegreat1 commented:
It seems like the best solutions are always narrate the original message or don't narrate modified messages. Do you have any suggestions?

Technici4n commented:
Hmmm, both options are not ideal. There's also the option to remove modification.

kevinthegreat1 commented:
Is it fine to just leave it there and maybe add a warning to mods? The modification feature doesn't affect anything when it's not used.

apple502j commented:
Modification is a security-sensitive part of the API, and I still believe the only way to do it is to cancel and send as system message.

I don't remember the specific details on how chat reporting works in 1.19.3, but it should be still true that clients send a message acknowledgment to the server, which constitute part of "evidences" signed and sent in a chat report. If the user sees a message modified by the client, the user didn't acknowledge the original message.

Sure, we could do what others are doing (see QuiltMC/quilt-standard-libraries#250) and implement a big chat API that is capable of doing everything. It might have made sense in 1.18, but now that the system is more complicated, exposing a bigger API surface is probably more of a maintenance burden than a benefit, especially here where an update is released within hours of snapshot release. There are not that much uses in client-side modifications as far as I can see. Plus, when an API is too costly to maintain here but is still useful, there's still an option to release it as a third-party API (like with Terrablender).

If we're keeping the modification functionality, it should just be a wrapper of cancellation event; if a message is modified, cancel and add ourselves. (Also, should this modify the original message or the decorated message? I prefer the former one, and perhaps using `Pair` of `Text` and `MessageType` so that narration works properly.)

kevinthegreat1 committed:
Removed modifying receive chat message

kevinthegreat1 commented:
Yes. I agree with the potential security issues related to users not actually seeing the message and the maintenance cost. I did not think of this. I am removing modification for received chat messages. There is now a message in the docs directing mods to use the allow event and add modified messages to the hud manually.

Looking at the quilt implementation, it seems like they're not dealing with all the char reporting. They're modifying the packet when it gets received. It will probably get handled the same way as chat messages modified by the server.

Technici4n commented on fabric-message-api-v1/src/client/java/net/fabricmc/fabric/api/client/message/v1/ClientReceiveMessageEvents.java:
```diff
@@ -0,0 +1,227 @@
+	 * An event triggered when the client receives a chat message,
+	 * which is any message sent by the server. Is not called when
+	 * {@linkplain #ALLOW_CHAT chat messages are blocked}.
+	 * Mods can use this to listen or modify the message.
```
This needs to be split into separate `MODIFY_GAME` and `GAME` events for modification and notification respectively.

kevinthegreat1 replied:
Should I split send message and command events to notify and modify too?

Technici4n replied:
Yeah.

kevinthegreat1 committed:
Split notify and modify events

kevinthegreat1 commented:
There are separate events for listening and modifying now.

kevinthegreat1 committed:
Fixed checkstyle

Technici4n reviewed (approved):
Works well, thank you!
